### ISSUE TITLE
ledger-api server: allow colon ':' in user ids

### ISSUE THREAD

**Issue by meiersi-da** [digital-asset/daml#12520]
As requested by @da-tanabe to support Kubernetes-provided names.

**Comment by da-tanabe** [comment]
TL;DR:
* colons are the only non-ASCII-alphanumeric characters that will appear in a Kubernetes subject _that originates from Kubernetes itself_
* no one seems to _really_ care about length or character restrictions in an `aud` field, and I think that's pretty odd.  :-)

### Kubernetes

Unfortunately the [Kubernetes documentation](https://kubernetes.io/docs/reference/access-authn-authz/rbac/#referring-to-subjects) is not very precise in its definition of subjects:

> The prefix `system:` is reserved for Kubernetes system use, so you should ensure that you don't have users or groups with names that start with `system:` by accident. Other than this special prefix, the RBAC authorization system does not require any format for usernames.

Kubernetes _does_ enforce additional rules on service account names (these comprise the components that make up a Kubernetes identity) that are not part of the documentation, but as the above link mentions, it does _not_ enforce any rules whatsoever on identities that are external to the cluster. An example of this is the identity that a human operator uses when interacting with a cluster.

* [func ValidateRoleBindingSubject in pkgapis/rbac/validation/validation.go](https://github.com/kubernetes/kubernetes/blob/f576fa9164e1cc25d8472ca02c945df936730bce/pkg/apis/rbac/validation/validation.go#L210-L249)
    * [func ValidateServiceAccountName in validation/generic.go](https://github.com/kubernetes/apimachinery/blob/162a22fc9219427f5fb9cab7a4a3a05bacc11f69/pkg/api/validation/generic.go#L65-L68)
    * [func IsDNS1123Subdomain in pkg/util/validation/validation.go](https://github.com/kubernetes/apimachinery/blob/162a22fc9219427f5fb9cab7a4a3a05bacc11f69/pkg/util/validation/validation.go#L207-L218), ultimately backed by the regular expression `^[a-z0-9]([-a-z0-9]*[a-z0-9])?$` (see [here](https://github.com/kubernetes/apimachinery/blob/162a22fc9219427f5fb9cab7a4a3a05bacc11f69/pkg/util/validation/validation.go#L178))

### JWT spec

I dug deeper into the [JWT spec](https://datatracker.ietf.org/doc/html/rfc7519) itself, specifically what _it_ defines as a valid string in a `sub` field ([4.1.2](https://datatracker.ietf.org/doc/html/rfc7519#section-4.1.2)). It is defined as a `StringOrURI`:

> A JSON string value, with the additional requirement that while arbitrary string values MAY be used, any value containing a ":" character MUST be a URI [[RFC3986](https://datatracker.ietf.org/doc/html/rfc3986)].  StringOrURI values are compared as case-sensitive strings with no transformations or canonicalizations applied.

Kubernetes subjects that refer to Kubernetes service accounts _do_ include colons, and luckily those conform to the URI spec.

**Comment by meiersi-da** [comment]
Thank you for the detailed analysis @da-tanabe . Given that checking whether a string is a URI is non-trivial (AFAIK), and that the IAM or particpiant operator is the gatekeeper of user name allocation, I'll just enable support for `:` in the usernames akin to how all other symbols are supported.
